bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java

```diff
@@ -43,54 +46,102 @@
   private static final long serialVersionUID = 2435897354284600685L;
 
   /**
-   * Converts a {@link CloudBigtableTableConfiguration} object to a
-   * {@link CloudBigtableScanConfiguration} that will perform the specified {@link Scan} on the
-   * table.
+   * Converts a {@link CloudBigtableTableConfiguration} object to a {@link
+   * CloudBigtableScanConfiguration} that will perform the specified {@link Scan} on the table.
+   *
    * @param config The {@link CloudBigtableTableConfiguration} object.
    * @param scan The {@link Scan} to add to the configuration.
    * @return The new {@link CloudBigtableScanConfiguration}.
    */
-  public static CloudBigtableScanConfiguration fromConfig(CloudBigtableTableConfiguration config,
-      Scan scan) {
+  public static CloudBigtableScanConfiguration fromConfig(
+      CloudBigtableTableConfiguration config, Scan scan) {
     CloudBigtableScanConfiguration.Builder builder = new CloudBigtableScanConfiguration.Builder();
     config.copyConfig(builder);
     return builder.withScan(scan).build();
   }
 
+  /**
+   * A {@link SerializableFunction} that creates a {@link ReadRowsRequest} from a {@link Scan}. This
+   * cannot be a anonymous class of {@link CloudBigtableScanConfiguration.Builder} because it would
+   * require the builder class to be serializable.
+   */
+  private static class ScanToRequestAdapter
+      implements SerializableFunction<SerializableScan, ReadRowsRequest> {
+    @Override
+    public ReadRowsRequest apply(SerializableScan scan) {
+      ReadHooks readHooks = new DefaultReadHooks();
+      ReadRowsRequest.Builder builder = Adapters.SCAN_ADAPTER.adapt(scan.get(), readHooks);
+      return readHooks.applyPreSendHook(builder.build());
+    }
+  }
+
   /**
    * Builds a {@link CloudBigtableScanConfiguration}.
    */
   public static class Builder extends CloudBigtableTableConfiguration.Builder {
-    private Scan scan;
-    private ReadRowsRequest request;
+    private ValueProvider<SerializableScan> scan;
+    private ValueProvider<ReadRowsRequest> request;
 
     public Builder() {
     }
 
     /**
-     * Specifies the {@link Scan} that will be used to filter the table.
-     * @param scan The {@link Scan} to add to the configuration.
+     * Specifies the {@link SerializableScan} that will be used to filter the table.
+     * TODO(kevinsi4508): We may not want pass {@link SerializableScan} runtime parameter because it
+     * is not easy to construct.
+     *
+     * @param scan The {@link SerializableScan} to add to the configuration.
      * @return The {@link CloudBigtableScanConfiguration.Builder} for chaining convenience.
      */
-    public Builder withScan(Scan scan) {
+    public Builder withScan(ValueProvider<SerializableScan> scan) {
```

Comment:
We should have a `ValueProvider<RowSet>` and `ValueProvider<RowFilter>`.  We should not expose `withScan<SerializableScan>`

Reply:
The purpose of this PR is templatizing the current config. We have more work to do with regard to replacing Scan but I believe we should deal with it later.

By templatizing the parameters first (with no regression), we can figure out what we should do next by trying out the template. we can discover what can't be runtime parameters, etc.

This may not be a good way to start creating template for cbt-sequenceFile work. If you have a better approach, please let me know.

Reply:
You introduced `SerializableScan` as a public interface.  I really don't like that, and would strongly prefer to have Bigtable protos in `ValueProviders` rather than having a `SerializableScan`.

Reply:
Lets deal with removing Scan first, then we can come back to this PR.